### PR TITLE
fix: allow IN queries on __key__

### LINKED ISSUE #1024: Query filter on __key__ using the "IN" operator
#### Context

The datastore API allows filtering on `__key__` with the "IN" operator and an array of Keys as value.  
I want to do using this library.

#### Problem description

This lib does not allow to query entities filtering on __key__ using the "IN" operator, and throws:  
`InvalidKey: A key should contain at least a kind.`  
=> The code expects a Key when I send an array of Keys.

#### Steps to reproduce

```js
const entitiesKey = [
  datastore.key({
    namespace,
    path: [kind, '01ece338-066d-443b-8c33-15414f971092']
  }),
  datastore.key({
    namespace,
    path: [kind, '0750f7ed-caf5-44d4-9105-0d72a81eeb9b']
  })
]
const query = datastore.createQuery(namespace, kind)
query.filter('__key__', 'IN', entitiesKey)
const entities = datastore.runQuery(query)
```

Throwing this:
```log
InvalidKey: A key should contain at least a kind.
    at Object.keyToKeyProto (C:\www\senfab\node_modules\@google-cloud\datastore\build\src\entity.js:974:19)
    at C:\www\senfab\node_modules\@google-cloud\datastore\build\src\entity.js:1108:45
    at Array.map (<anonymous>)
    at Object.queryToQueryProto (C:\www\senfab\node_modules\@google-cloud\datastore\build\src\entity.js:1104:43)
    at makeRequest (C:\www\senfab\node_modules\@google-cloud\datastore\build\src\request.js:273:49)
```

#### Results explanation

=> In build/src/entity.js l.1107, in the queryToQueryProto function, these lines are causing me trouble:
````js
if (filter.name === '__key__') {
    value.keyValue = entity.keyToKeyProto(filter.val);
}
else {
    value = entity.encodeValue(filter.val, filter.name);
}
````
=> `entity.encodeValue` handle correctly if `filter.val` is an array, but `entity.keyToKeyProto` doesn't handle arrays,  
whereas the API does handle filtering on `__key__` with the "IN" operator and an array as value.

#### Code suggestion

Patching this code is enough to make it work:
```js
if (filter.name === '__key__') {
  if (Array.isArray(filter.val)) {
    value = {
      arrayValue: {
        values: filter.val.map(v => {
          return {
            keyValue: entity.keyToKeyProto(v)
          }
        })
      }
    }
  } else {
    value.keyValue = entity.keyToKeyProto(filter.val)
  }
} else {
  value = entity.encodeValue(filter.val, filter.name)
}
```

#### Environment details

  - OS: Windows 10
  - Node.js version: 16.14.0
  - npm version: 8.3.1
  - `@google-cloud/datastore` version: 7.0.0

### PR DESCRIPTION
**Pull request by danieljbruce**
The fix contains:

- An update to the way that entities are encoded
- An update to tests so that an instanceOf check passes
- A cleanup of the PropertyFilter function
- An additional test to check for IN with keys

Fixes #1024

### CODE REVIEW

File: src/filter.ts [hunk 1]
```diff
@@ -13,7 +13,7 @@
 // limitations under the License.
 
 import {Operator, Filter as IFilter} from './query';
-import {entity} from './entity';
+import {entity, ValueProto} from './entity';
```
Comment (vicb) [inline src/filter.ts:16]: Please fix (remove `ValueProto`)

File: src/filter.ts [hunk 2]
```diff
@@ -100,11 +81,7 @@ export class PropertyFilter extends EntityFilter implements IFilter {
    */
   // eslint-disable-next-line
   toProto(): any {
-    const value = new PropertyFilter(
-      this.name,
-      this.op,
-      this.val
-    ).encodedValue();
+    const value = entity.encodeValue(this.val, this.name);
```
Comment (vicb) [inline src/filter.ts:84]: You can also `value: entity.encodeValue(this.val, this.name)` inline. Not sure what the preferred style is.

**Review by vicb — APPROVED** [review]
Looks nice 👍 

There are a few "GitHub Actions / lint" that look worth addressing - IMO the typings of this lib could use some love but out of scope for this PR.

File: test/entity.ts [hunk 3]
```diff
@@ -1894,8 +1894,62 @@ describe('entity', () => {
       assert.deepStrictEqual(entity.queryToQueryProto(query), queryProto);
     });
 
+    it('should support using key with IN', () => {
```
[the remark this answers is not in the document]
Comment (danieljbruce) [reply test/entity.ts:1897]: Good point.

I think more specifically, it is addressing the case of using an array as a value as that was what was causing the error in the first place.

File: src/filter.ts (hunk 2 above)
Comment (danieljbruce) [reply src/filter.ts:84]: Yes. This can be even more compact.

File: src/filter.ts (hunk 1 above)
Comment (danieljbruce) [reply src/filter.ts]: Done

**Comment by danieljbruce** [comment]
@vicb There is just one more thing we should confirm. I notice that `val` in Property Filter is of type `any` so even though filtering by `__key__` ought to use only values that are keys and arrays of keys, the compiler doesn't guarantee it. Are these definitely the only two cases with which we can expect users to filter with `__key__`?

File: src/filter.ts [hunk 4]
```diff
@@ -56,42 +57,34 @@ export abstract class EntityFilter {
   abstract toProto(): any;
 }
 
+export type restrictedValueType<T> = T extends '__key__'
```
Comment (vicb) [inline src/filter.ts:60]: `RestrictedValueType`

File: src/query.ts [hunk 5]
```diff
@@ -207,12 +207,19 @@ class Query {
...
-  filter(propertyOrFilter: string | EntityFilter, value?: {} | null): Query;
-  filter(propertyOrFilter: string, operator: Operator, value: {} | null): Query;
-  filter(
-    propertyOrFilter: string | EntityFilter,
+  filter<T extends string>(
+    propertyOrFilter: T | EntityFilter,
+    value?: restrictedValueType<T> | null
+  ): Query;
+  filter<T extends string>(
+    propertyOrFilter: T,
```
Comment (vicb) [inline src/query.ts:215]: T | EntityFilter

File: src/query.ts [hunk 6]
```diff
@@ -207,12 +207,19 @@ class Query {
...
+    operator: Operator,
+    value: restrictedValueType<T> | null
+  ): Query;
+  filter<T extends string>(
+    propertyOrFilter: T | EntityFilter,
     operatorOrValue?: Operator,
```
Comment (vicb) [inline src/query.ts:221]: This should be `restrictedValueType<T> | null | Operator`

**Comment by danieljbruce** [comment]
I mean see https://github.com/googleapis/nodejs-datastore/pull/1089/commits/8d269d2e2cd60bef27ff6856bffe21c5a8f9f20c

**Comment by danieljbruce** [comment]
I mean https://github.com/googleapis/nodejs-datastore/pull/1089

File: src/query.ts (hunk 6 above)
Comment (danieljbruce) [reply src/query.ts]: Signature has been updated

File: src/filter.ts [hunk 7]
```diff
@@ -156,5 +144,5 @@ class CompositeFilter extends EntityFilter {
 }
 
 export function isFilter(filter: any): filter is EntityFilter {
-  return (filter as EntityFilter).toProto !== undefined;
+  return filter instanceof EntityFilter;
```
Comment (kolea2) [inline src/filter.ts:147]: why does this have to change?

Comment (vicb) [reply src/filter.ts:147]: I suggested the fix so I'll answer.

`filter.toProto !== undefined` only checks if the object has a `toProto` properties (not set to `undefined`)
It is the case that `EntityFilter`s have a `toProto` methods but it is not necessarily the case that an object with a `toProto` property is an `EntityFilter`.

`filter instanceof EntityFilter` is the correct check for `filter is EntityFilter`

I hope that makes sense.

File: src/query.ts [hunk 8]
```diff
@@ -207,12 +207,20 @@ class Query {
    * const keyQuery = query.filter('__key__', key);
    * ```
    */
-  filter(propertyOrFilter: string | EntityFilter, value?: {} | null): Query;
-  filter(propertyOrFilter: string, operator: Operator, value: {} | null): Query;
-  filter(
-    propertyOrFilter: string | EntityFilter,
-    operatorOrValue?: Operator,
-    value?: {} | null
+  filter(filter: EntityFilter): Query;
```
Comment (danieljbruce) [inline src/query.ts:210]: This signature change makes the input parameters more restrictive. Do we really want to go through with this? If so would this be a major version bump?

@vicb @bcoe

Comment (vicb) [reply src/query.ts:210]: My take is that the old signature was buggy and wouldn't catch unsupported calls. So I see this as a bug fix.

Also I think the impl could be improved by `switch`ing on `arguments.length`. 1, 2, 3 corresponding to the first 3 signatures in order.

Comment (danieljbruce) [reply src/query.ts:210]: I will do this in a separate issue:

https://github.com/googleapis/nodejs-datastore/issues/1096

**Review by ruyadorno — APPROVED** [review]
LGTM 👍 

Let's just make sure to keep an eye and monitor if the internal change from `src/filter.ts` `encodeValue()` is going to have any impact in the end users of the library. Releasing this as a patch version release is going to help us having the flexibility to rollback if needed, so I'd advise to proceed with a `fix:` commit.

Great work @danieljbruce!

**Pull request merged**